File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False

        char_to_words = {}
        words_to_char = {}

        for ch, word in zip(pattern, words):

            if ch in char_to_words and char_to_words[ch] != word:
               return False

            if word in words_to_char and words_to_char[word] != ch:
               return False         


            char_to_words[ch] = word
            words_to_char[word] = ch

        return True    

File: 0290-word-pattern/0290-word-pattern.dart
class Solution {
  bool wordPattern(String pattern, String s) {
       List<String> words = s.split(" ");

       if(words.length != pattern.length) return false;

       Map<String, int> charIndex = {};
       Map<String, int> wordIndex = {};

       for(int i = 0 ; i < pattern.length ; i++) {
           
           if(charIndex[pattern[i]] != wordIndex[words[i]]) return false;

           charIndex[pattern[i]] = i;
           wordIndex[words[i]] = i;
       }
    return true;   
  }
}